### ISSUE TITLE
WaitAndRetry for given time rather than retry count

### ISSUE THREAD

**Issue by sanjaysingh** [App-vNext/Polly#327]
I need to retry a request for 2 hours and then give up. I will make the request with the exponential back off of 2 seconds with maximum wait for 5 minutes.  I don't see any API to allow for retry timespan rather than retrycount. I can do that by generating the timespan array or doing something as follows where i figure out the count that will take me close to 2 hours.

Policy
                        .Handle<Exception>()
                        .WaitAndRetry(25,
                        retryAttempt =>
                        {
                            var waiTimeSpan = TimeSpan.FromSeconds(Math.Pow(2, retryAttempt));
                            if (waiTimeSpan.TotalMinutes > 5)
                            {
                                waiTimeSpan = TimeSpan.FromMinutes(5);
                            }
                            return waiTimeSpan;
                        },
                        (exception, timeSpan, retryAttempt, context) =>
                        {
                           // log the exception
                        });

**Comment by reisenberger** [comment]
Hi @sanjaysingh .  You can use [`PolicyWrap`](https://github.com/App-vNext/Polly/wiki/PolicyWrap) to wrap a [`TimeoutPolicy`](https://github.com/App-vNext/Polly/wiki/Timeout) around the `WaitAndRetry(...)` policy.  [This similar question](https://github.com/App-vNext/Polly/issues/261) describes in more detail how to do this, and links to sample code in DotNetFiddle.

Note that the _synchronous_ `WaitAndRetry(...)` blocks the thread while waiting. Polly isn't a 'background job scheduler' that schedules work for a later time. Polly's WaitAndRetry() is primarily designed for retrying transient faults after a short interval.  If you are intending long waits on synchronous code paths, at least especially if you might have multiple threads doing this, you might also consider `Timer`-driven callbacks or (if a web app) a background-work scheduler like [hangfire](https://www.hangfire.io/), so as not to block threads.  You could still use the Polly Policy within the scheduled (hangfire or similar) task  to make the shorter-interval retries (eg after 2, 4, 8 etc up to 60 (or whatever you like) seconds), then switch to scheduling a new task with a work-scheduler, for the longer pauses.  (That may or may not be relevant depending on your requirements - but in case useful.)

In an `async`/`await` context, Polly's `WaitAndRetryAsync(...)` policy waits using `Task.Delay(...)`.

**Comment by sanjaysingh** [comment]
Thank you @reisenberger PolicyWrap is awesome and I am now able to achieve this. I understand that the WaitAndRetry is synchronous and that is ok for my case because i have those as backgroound Tasks which each run in isolation with their own retries. Thank you for your help.

**Comment by reisenberger** [comment]
Perfect. Glad it helped.
